widgets/src/LayoutIndicator.c

```diff
@@ -77,11 +78,7 @@ struct _AnacondaLayoutIndicatorPrivate {
     GtkWidget *icon;
     GtkLabel *layout_label;
     GdkCursor *cursor;
-    XklConfigRec *config_rec;
-    gulong state_changed_handler_id;
-    gboolean state_changed_handler_id_set;
-    gulong config_changed_handler_id;
-    gboolean config_changed_handler_id_set;
+    GkKioskInputSources *gk_proxy;
```

Comment:
Couldn't we change this to remove the logic from C code and move it to python only. I mean leaving this widget only as `view` for the python code.

I mean do not connect this to Kiosk proxy but just leave the python code which creates this component to communicate with the python and set / get / signal values based on the user actions to python code.

That would simplify the future transition to something else a lot.

Reply:
Yes, that should be possible. Where is the right place to move the service? `pyanaconda/modules/localization`? It seems like other keyboard related code is there.

cc @M4rtinK

Reply:
@jexposit Yeah, I think that's the right place for such a thing. Though I think what @jkonecny12 suggested was to maximally simplify the keyboard switching widget, so that it essentially provides just a "view" & the logic would be in the part of the GUI code implemented in Python.

Reply:
Basically, the view could provide just these 3 things:
* a method to set the displayed keyboard layout value
* a method to read the displayed keyboard value (this might actually not be 100% necessary)
* a signal that is triggered when the widget is clicked
That way the keyboard layout switching logic could be implemented in Python, which could be easier for us to maintain.

Still, I think we will need an Anaconda DBus API for switching the keyboards, as its not just the UI but also the backend (via kickstart or boot options) that does keyboard switching & it does not make much sense to implement it twice, once in GUI & once in the backend.

Reply:
Let me try to summarize what I mean via not-very-fancy ASCII diagram

Keyboard layout set via kickstart or boot option
---------------------------------------------------------------------

`kickstart/boot-option -> backend localization module uses appropriate API (localed DBus/GNOME KIOSK DBus)`

Keyboard layout set via GUI - GTK/C widget using DBus directly
--------------------------------------------------------------------------------------------

`widget is clicked -> calls Anaconda DBus API for setting keyboard layout -> backend localization module uses appropriate API (localed DBus/GNOME KIOSK DBus)`


Keyboard layout set via GUI - GTK/C widget with logic handled by Python code
-----------------------------------------------------------------------------------------------------------------

`widget is clicked -> triggers a signal that Python GUI code is connected to -> Python GUI code calls Anaconda DBus API for setting keyboard layout -> backend localization module uses appropriate API (localed DBus API/GNOME KIOSK DBus API)`

Looking at it this way, I would say both cases might come of more or less the same complexity wise, as we will need the Anaconda provided DBus API anyway, so if its called from C or Python code might not make so much difference.

Reply:
Thanks for the details Martin. 5 new methods in the localization service:

- get_compositor_selected_layout
- set_compositor_selected_layout
- get_compositor_layouts
- set_compositor_layouts
- select_next_compositor_layout

Should be enough to handle all our needs. 

I'll do some testing with option 3. I think that adding the logic in `pyanaconda/ui/gui/__init__.py` will cover all cases. However, since I see another instance of `AnacondaLayoutIndicator` in `passphrase.glade`, I might use Anaconda's D-Bus service directly from `LayoutIndicator.c` to avoid duplicating code.

Reply:
That sounds good to me. Thanks!

Reply:
The `LayoutIndicator` C widget is now using Anaconda's localization module over D-Bus.

If in the future the backend is changed from GNOME Kiosk to something else, it could be done without changing the C code.